File: Kernel Simulator/Misc/Calendar/CalendarPrint.vb

'    Kernel Simulator  Copyright (C) 2018-2022  Aptivi
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports KS.Misc.Calendar.Events
Imports KS.Misc.Calendar.Reminders
Imports KS.Misc.Writers.FancyWriters.Tools
Imports Terminaux.Writer.FancyWriters.Tools

Namespace Misc.Calendar
    Public Module CalendarPrint

        ''' <summary>
        ''' Prints the table of the calendar
        ''' </summary>
        Public Sub PrintCalendar()
            PrintCalendar(Date.Today.Year, Date.Today.Month)
        End Sub

        ''' <summary>
        ''' Prints the table of the calendar
        ''' </summary>
        Public Sub PrintCalendar(Year As Integer, Month As Integer)
            Dim CalendarDays As String() = CurrentCult.DateTimeFormat.DayNames
            Dim CalendarMonths As String() = CurrentCult.DateTimeFormat.MonthNames
            Dim CalendarData(5, CalendarDays.Length - 1) As String
            Dim DateFrom As New Date(Year, Month, 1, CurrentCult.Calendar)
            Dim DateTo As New Date(Year, Month, Date.DaysInMonth(Year, Month), CurrentCult.Calendar)
            Dim CurrentWeek As Integer = 1
            Dim CalendarTitle As String = CalendarMonths(Month - 1) & " " & Year
            Dim CalendarCellOptions As New List(Of CellOptions)

            'Populate the calendar data
            WriteWhere(CalendarTitle, CInt((ConsoleWrapper.WindowWidth - CalendarTitle.Length) / 2), ConsoleWrapper.CursorTop, True, GetConsoleColor(ColTypes.Neutral))
            WritePlain("", True)
            For CurrentDay As Integer = 1 To DateTo.Day
                Dim CurrentDate As New Date(Year, Month, CurrentDay, CurrentCult.DateTimeFormat.Calendar)
                If CurrentDate.DayOfWeek = 0 Then CurrentWeek += 1
                Dim CurrentWeekIndex As Integer = CurrentWeek - 1
                Dim CurrentDayMark As String = $" {CurrentDay} "
                Dim ReminderMarked As Boolean = False
                Dim EventMarked As Boolean = False
                Dim IsWeekend As Boolean = CurrentDate.DayOfWeek = DayOfWeek.Friday Or CurrentDate.DayOfWeek = DayOfWeek.Saturday

                'Dim out the weekends
                If IsWeekend Then
                    Dim WeekendOptions As New CellOptions(CurrentDate.DayOfWeek + 1, CurrentWeek) With {
                    .ColoredCell = True,
                    .CellColor = New Color(128, 128, 128),
                    .CellBackgroundColor = BackgroundColor
                }
                    CalendarCellOptions.Add(WeekendOptions)
                End If

                'Know where and how to put the day number
                For Each Reminder As ReminderInfo In Reminders.Reminders
                    If Reminder.ReminderDate.Date = CurrentDate And Not ReminderMarked Then
                        CurrentDayMark = $"({CurrentDay})"
                        ReminderMarked = True
                    End If
                Next
                For Each EventInstance As EventInfo In CalendarEvents
                    If EventInstance.EventDate = CurrentDate And Not EventMarked Then
                        Dim EventCell As New CellOptions(CurrentDate.DayOfWeek + 1, CurrentWeek) With {
                        .ColoredCell = True,
                        .CellColor = StageColor,
                        .CellBackgroundColor = BackgroundColor
                    }
                        CalendarCellOptions.Add(EventCell)
                        EventMarked = True
                    End If
                Next
                CalendarData(CurrentWeekIndex, CurrentDate.DayOfWeek) = CurrentDayMark
            Next
            WriteTable(CalendarDays, CalendarData, 2, True, CalendarCellOptions)
        End Sub

    End Module
End Namespace

File: Kernel Simulator/Misc/Calendar/CalendarPrint.cs
﻿//
// Kernel Simulator  Copyright (C) 2018-2024  Aptivi
//
// This file is part of Kernel Simulator
//
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY, without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

using System;
using System.Collections.Generic;
using KS.ConsoleBase.Colors;
using KS.Languages;
using KS.Misc.Calendar.Events;
using KS.Misc.Calendar.Reminders;
using KS.ConsoleBase.Writers;
using Terminaux.Base;
using Terminaux.Colors;
using Terminaux.Writer.FancyWriters.Tools;

namespace KS.Misc.Calendar
{
    public static class CalendarPrint
    {

        /// <summary>
        /// Prints the table of the calendar
        /// </summary>
        public static void PrintCalendar()
        {
            PrintCalendar(DateTime.Today.Year, DateTime.Today.Month);
        }

        /// <summary>
        /// Prints the table of the calendar
        /// </summary>
        public static void PrintCalendar(int Year, int Month)
        {
            string[] CalendarDays = CultureManager.CurrentCult.DateTimeFormat.DayNames;
            string[] CalendarMonths = CultureManager.CurrentCult.DateTimeFormat.MonthNames;
            var CalendarData = new string[6, CalendarDays.Length];
            _ = new DateTime(Year, Month, 1, CultureManager.CurrentCult.Calendar);
            var DateTo = new DateTime(Year, Month, DateTime.DaysInMonth(Year, Month), CultureManager.CurrentCult.Calendar);
            int CurrentWeek = 1;
            string CalendarTitle = CalendarMonths[Month - 1] + " " + Year;
            var CalendarCellOptions = new List<CellOptions>();

            // Populate the calendar data
            TextWriters.WriteWhere(CalendarTitle, (int)Math.Round((ConsoleWrapper.WindowWidth - CalendarTitle.Length) / 2d), ConsoleWrapper.CursorTop, true, KernelColorTools.ColTypes.Neutral);
            TextWriters.Write("", KernelColorTools.ColTypes.Neutral);
            for (int CurrentDay = 1, loopTo = DateTo.Day; CurrentDay <= loopTo; CurrentDay++)
            {
                var CurrentDate = new DateTime(Year, Month, CurrentDay, CultureManager.CurrentCult.DateTimeFormat.Calendar);
                if (CurrentDate.DayOfWeek == 0)
                    CurrentWeek += 1;
                int CurrentWeekIndex = CurrentWeek - 1;
                string CurrentDayMark = $" {CurrentDay} ";
                bool ReminderMarked = false;
                bool EventMarked = false;
                bool IsWeekend = CurrentDate.DayOfWeek == DayOfWeek.Friday | CurrentDate.DayOfWeek == DayOfWeek.Saturday;

                // Dim out the weekends
                if (IsWeekend)
                {
                    var WeekendOptions = new CellOptions((int)CurrentDate.DayOfWeek + 1, CurrentWeek)
                    {
                        ColoredCell = true,
                        CellColor = new Color(128, 128, 128),
                        CellBackgroundColor = KernelColorTools.BackgroundColor
                    };
                    CalendarCellOptions.Add(WeekendOptions);
                }

                // Know where and how to put the day number
                foreach (ReminderInfo Reminder in ReminderManager.Reminders)
                {
                    if (Reminder.ReminderDate.Date == CurrentDate & !ReminderMarked)
                    {
                        CurrentDayMark = $"({CurrentDay})";
                        ReminderMarked = true;
                    }
                }
                foreach (EventInfo EventInstance in EventManager.CalendarEvents)
                {
                    if (EventInstance.EventDate == CurrentDate & !EventMarked)
                    {
                        var EventCell = new CellOptions((int)CurrentDate.DayOfWeek + 1, CurrentWeek)
                        {
                            ColoredCell = true,
                            CellColor = KernelColorTools.StageColor,
                            CellBackgroundColor = KernelColorTools.BackgroundColor
                        };
                        CalendarCellOptions.Add(EventCell);
                        EventMarked = true;
                    }
                }
                CalendarData[CurrentWeekIndex, (int)CurrentDate.DayOfWeek] = CurrentDayMark;
            }
            TextFancyWriters.WriteTable(CalendarDays, CalendarData, 2, KernelColorTools.ColTypes.Neutral, KernelColorTools.ColTypes.Neutral, KernelColorTools.ColTypes.Neutral, KernelColorTools.ColTypes.Neutral, true, CalendarCellOptions);
        }

    }
}